File: Java/VKfragment/app/src/main/java/com/example/vkfragment/HomeFragment.java
package com.example.vkfragment;

import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.vkfragment.models.DataNews;
import com.example.vkfragment.models.News;

import java.util.ArrayList;
import java.util.List;

public class HomeFragment extends Fragment {
    //container for all news from database
    public List<News> items = new ArrayList<>();

    private RecyclerView recyclerView;
    private NewsAdapter adapter;
    private NewsAdapter.ItemClickListener listener = null;
    private final String STATE = "recycler_state";
    private static Bundle recyclerviewState;

    public HomeFragment(){
    }


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_home, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        listener = new NewsAdapter.ItemClickListener() {
            @Override
            public void itemClick(int position, News item) {
                Intent intent = new Intent(getActivity(),NewsDetailActivity.class);
                intent.putExtra("news", item);
                startActivity(intent);
            }
        };

        recyclerView = view.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(view.getContext()));
        adapter = new NewsAdapter(generateNews(), listener);
        recyclerView.setAdapter(adapter);

    }
    //return data from newsDATABASE
    private List<News> generateNews() {
        items = DataNews.news;
        return items;
    }

    @Override
    public void onResume(){
        super.onResume();
        if (recyclerviewState != null){
            Parcelable listState = recyclerviewState.getParcelable(STATE);
            recyclerView.getLayoutManager().onRestoreInstanceState(listState);
        }
    }

    @Override
    public void onPause(){
        super.onPause();
        recyclerviewState = new Bundle();
        Parcelable listState = recyclerView.getLayoutManager().onSaveInstanceState();
        recyclerviewState.putParcelable(STATE, listState);
    }
}

File: Kotlin/VKfragmentKotlin/app/src/main/java/com/example/vkfragmentkotlin/HomeFragment.kt
package com.example.vkfragmentkotlin

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.vkfragmentkotlin.models.DataNews
import com.example.vkfragmentkotlin.models.News

class HomeFragment : Fragment(), NewsAdapter.RecyclerViewItemClick  {

    private lateinit var recyclerView: RecyclerView
    private lateinit var newsAdapter: NewsAdapter

    private val STATE = "recycler_view"
    private var recyclerViewState: Bundle?=null

    lateinit var  items : List<News>

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
         return  inflater.inflate(R.layout.fragment_home,container,false)
    }

    //click to intent for details of news
    override fun itemClick(position: Int, item: News)  {
        val intent = Intent(activity, NewsDetailActivity::class.java)
        intent.putExtra("news", item)
        startActivity(intent)
    }

    //init of recyclerview and adapter
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recyclerView = activity!!.findViewById(R.id.recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(activity)
        newsAdapter = NewsAdapter(GenerateNews(),itemClickListener = this)
        recyclerView.adapter = newsAdapter
    }

    //overriding fun to restore saved state of recyclerview
    override fun onResume() {
        super.onResume()
        if (recyclerViewState!=null){
            val listState = recyclerViewState!!.getParcelable<Parcelable>(STATE)
            recyclerView.layoutManager!!.onRestoreInstanceState(listState)
        }
    }

    //overriding fun to save the state of recyclerview news
    override fun onPause() {
        super.onPause()
        recyclerViewState = Bundle()
        val listState = recyclerView.layoutManager!!.onSaveInstanceState()
        recyclerViewState!!.putParcelable(STATE, listState)
    }

    //fetching news list from dataNews object
    fun GenerateNews() : List<News>  {
       items = DataNews.newsBIG
        return items
    }
}